src/mapport.cpp

```diff
@@ -0,0 +1,320 @@
+            } else {
+                if (externalIPAddress[0]) {
+                    CNetAddr resolved;
+                    if (LookupHost(externalIPAddress, resolved, false)) {
+                        LogPrintf("UPnP: ExternalIPAddress = %s\n", resolved.ToString());
+                        AddLocal(resolved, LOCAL_MAPPED);
+                    }
+                } else {
+                    LogPrintf("UPnP: GetExternalIPAddress failed.\n");
+                }
+            }
+        }
+
+        std::string strDesc = PACKAGE_NAME " " + FormatFullVersion();
+
+        do {
+            r = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, port.c_str(), port.c_str(), lanaddr, strDesc.c_str(), "TCP", 0, "0");
+
+            if (r != UPNPCOMMAND_SUCCESS) {
+                ret = false;
+                LogPrintf("AddPortMapping(%s, %s, %s) failed with code %d (%s)\n", port, port, lanaddr, r, strupnperror(r));
+                break;
+            } else {
+                ret = true;
+                g_mapport_current_proto = MapPortProto::UPNP;
+                LogPrintf("UPnP Port Mapping successful.\n");
+            }
+        } while (g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+
+        r = UPNP_DeletePortMapping(urls.controlURL, data.first.servicetype, port.c_str(), "TCP", 0);
+        LogPrintf("UPNP_DeletePortMapping() returned: %d\n", r);
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        FreeUPNPUrls(&urls);
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
+    }
+
+    return ret;
+}
+#endif // #ifdef USE_UPNP
+
+static void ThreadMapPort()
+{
+    bool ok;
+    do {
+        ok = false;
+
+#ifdef USE_NATPMP
+        if (g_mapport_target_proto & MapPortProto::NAT_PMP) {
+            ok = ThreadNatpmp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_NATPMP
+
+#ifdef USE_UPNP
+        if (g_mapport_target_proto & MapPortProto::UPNP) {
+            ok = ThreadUpnp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_UPNP
+
+        if (g_mapport_target_proto == MapPortProto::NONE) {
+            g_mapport_current_proto = MapPortProto::NONE;
+            return;
+        }
+
+    } while (ok || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD));
+}
+
+void StartMapPort(bool use_natpmp, bool use_upnp)
+{
+    if (use_natpmp) {
+        g_mapport_target_proto |= MapPortProto::NAT_PMP;
+    } else {
+        g_mapport_target_proto &= ~MapPortProto::NAT_PMP;
+    }
+
+    if (use_upnp) {
+        g_mapport_target_proto |= MapPortProto::UPNP;
+    } else {
+        g_mapport_target_proto &= ~MapPortProto::UPNP;
+    }
+
+    if (g_mapport_current_proto == MapPortProto::NONE) {
+        if (g_mapport_target_proto == MapPortProto::NONE) {
+            return;
+        }
+    } else if (g_mapport_target_proto & g_mapport_current_proto) {
```

Comment:
This won't work if we used to have both protocols enabled, are currently using UPnP, but want to disable UPnP.

Reply:
Sure it works.
The log excerpt for the mentioned case:
```
2020-02-20T23:26:30Z [mapport] UPNP_DeletePortMapping() returned: 0
2020-02-20T23:26:30Z [mapport] AddLocal(95.164.65.194:18333,3)
2020-02-20T23:26:30Z [mapport] NAT-PMP: Port mapping successful. External address = 95.164.65.194:18333

```

Reply:
Why would it? You'd return early here without making the change...

Reply:
If both protocols were enabled, `g_mapport_target_proto != MapPortProto::NONE`

Reply:
The conditional here is `(g_mapport_target_proto & g_mapport_current_proto)`

Reply:
```C++
void StartMapPort(bool use_natpmp, bool use_upnp)
{
    // Both protocols are enabled, so
    // g_mapport_target_proto == MapPortProto::NAT_PMP | MapPortProto::UPNP
    //
    // UPnP is currently using, so
    // g_mapport_current_proto == MapPortProto::UPNP
    //
    // A user disables UPnP then passed arguments are:
    // use_natpmp == true, use_upnp = false

    if (use_natpmp) {
        // Effectively noop here as MapPortProto::NAT_PMP is set already.
        g_mapport_target_proto |= MapPortProto::NAT_PMP;
    } else {
        g_mapport_target_proto &= ~MapPortProto::NAT_PMP;
    }

    if (use_upnp) {
        g_mapport_target_proto |= MapPortProto::UPNP;
    } else {
        // Unset MapPortProto::UPNP bit.
        g_mapport_target_proto &= ~MapPortProto::UPNP;
    }

    // At this point variables are set to the following values:
    // g_mapport_target_proto == MapPortProto::NAT_PMP
    // g_mapport_current_proto == MapPortProto::UPNP
    // ...
    if (g_mapport_current_proto == MapPortProto::NONE) {
        if (g_mapport_target_proto == MapPortProto::NONE) {
            return;
        }
    // ... and the next condition evaluates to false.
    } else if (g_mapport_target_proto & g_mapport_current_proto) {
        return;
    }

    // Making the changes...
    ...
}
```

And logs confirm this logic.

Reply:
Got it. I was confusing it with (target & new_target)